mlir/include/mlir/Analysis/Presburger/Barvinok.h

```diff
@@ -0,0 +1,237 @@
+             "parameters!")
+    }
+  }
+
+  // Find the number of parameters involved in the polynomial.
+  unsigned getNumParams() { return numParam; }
+
+  QuasiPolynomial operator+(const QuasiPolynomial &x) {
+    assert(numParam == x.getNumParams() &&
+           "two quasi-polynomials with different numbers of parameters cannot "
+           "be added!");
+    coefficients.append(x.coefficients);
+    affine.insert(affine.end(), x.affine.begin(), x.affine.end());
+    return *this;
+  }
+
+  QuasiPolynomial operator-(const QuasiPolynomial &x) {
+    bool sameNumParams = (getNumParams() == -1) || (x.getNumParams() == -1) ||
+                         (getNumParams() == x.getNumParams());
+    assert(numParam == x.getNumParams() &&
+           "two quasi-polynomials with different numbers of parameters cannot "
+           "be subtracted!");
+    QuasiPolynomial qp(x.coefficients, x.affine);
+    for (unsigned i = 0, e = x.coefficients.size(); i < e; i++)
+      qp.coefficients[i] = -qp.coefficients[i];
+    return (*this + qp);
+  }
+
+  QuasiPolynomial operator*(const QuasiPolynomial &x) {
+    assert(numParam = x.getNumParams() &&
+                      "two quasi-polynomials with different numbers of "
+                      "parameters cannot be multiplied!");
+    QuasiPolynomial qp();
+    std::vector<SmallVector<Fraction>> product;
+    for (unsigned i = 0, e = coefficients.size(); i < e; i++) {
+      for (unsigned j = 0, e = x.coefficients.size(); j < e; j++) {
+        qp.coefficients.append({coefficients[i] * x.coefficients[j]});
+
+        product.clear();
+        product.insert(product.end(), affine[i].begin(), affine[i].end());
+        product.insert(product.end(), x.affine[j].begin(), x.affine[j].end());
+
+        qp.affine.push_back(product);
+      }
+    }
+
+    return qp;
+  }
+
+  QuasiPolynomial operator/(Fraction x) {
+    assert(x != 0 && "division by zero!");
+    for (Fraction &coeff : coefficients)
+      coeff /= x;
+    return *this;
+  };
+
+  // Removes terms which evaluate to zero from the expression.
+  QuasiPolynomial simplify() {
+    SmallVector<Fraction> newCoeffs({});
+    std::vector<std::vector<SmallVector<Fraction>>> newAffine({});
+    bool prodIsNonz, sumIsNonz;
+    for (unsigned i = 0, e = coefficients.size(); i < e; i++) {
+      prodIsNonz = true;
+      for (unsigned j = 0, e = affine[i].size(); j < e; j++) {
+        sumIsNonz = false;
+        for (unsigned k = 0, e = affine[i][j].size(); k < e; k++)
+          if (affine[i][j][k] != Fraction(0, 1))
+            sumIsNonz = true;
+        if (sumIsNonz == false)
+          prodIsNonz = false;
+      }
+      if (prodIsNonz && coefficients[i] != Fraction(0, 1)) {
+        newCoeffs.append({coefficients[i]});
+        newAffine.push_back({affine[i]});
+      }
+    }
+    return QuasiPolynomial(newCoeffs, newAffine);
+  }
+
+  SmallVector<Fraction> coefficients;
+  std::vector<std::vector<SmallVector<Fraction>>> affine;
+
```

Comment:
Why are these public?

Reply:
They need to be set during the constant term computation

Reply:
Won't it just construct and return one of these? That should be possible from the public interface anyway

Reply:
Ah right. Will refactor